service/src/SysfsIOGroup.cpp

```diff
@@ -0,0 +1,550 @@
+
+                // TODO (dcw): Make sure parser is scaling for us
+                info.value = info.parse(std::string(info.buf.data()));
+            }
+        }
+    }
+
+    void SysfsIOGroup::write_batch(void)
+    {
+        m_is_batch_write = true;
+        if (!m_batch_writer) {
+            m_batch_writer = IOUring::make_unique(m_pushed_info_signal.size());
+        }
+
+        for (auto &info : m_pushed_info_control) {
+            if (info.do_write && !std::isnan(info.value)) {
+                std::string setting = info.gen(info.value);
+                if (setting.length() >= info.buf.size()) {
+                    throw geopm::Exception("SysfsIOGroup control value is too long",
+                                           GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+                }
+                std::strncpy(info.buf.data(), setting.c_str(), info.buf.size());
+                m_batch_writer->prep_write(
+                    info.last_io_return, info.fd.get(), info.buf.data(),
+                    static_cast<unsigned>(setting.length() + 1), 0);
+            }
+        }
+        m_batch_writer->submit();
+        for (auto &info : m_pushed_info_control) {
+            if (info.do_write && !std::isnan(info.value)) {
+                if (*info.last_io_return < 0) {
+                    throw geopm::Exception("SysfsIOGroup failed to write control \"" +
+                                           info.name + "\"",
+                                           errno, __FILE__, __LINE__);
+                }
+            }
+        }
+    }
+
+    double SysfsIOGroup::sample(int batch_idx)
+    {
+        if (batch_idx < 0 || static_cast<size_t>(batch_idx) >= m_pushed_info_signal.size()) {
+            throw Exception("SysfsIOGroup::sample(): batch_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (!m_is_batch_read) {
+            throw Exception("SysfsIOGroup::sample(): signal has not been read.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return m_pushed_info_signal[static_cast<size_t>(batch_idx)].value;
+    }
+
+    // Save a setting to be written by a future write_batch()
+    void SysfsIOGroup::adjust(int batch_idx, double setting)
+    {
+        if (batch_idx < 0 || static_cast<size_t>(batch_idx) >= m_pushed_info_control.size()) {
+            throw Exception("SysfsIOGroup::adjust(): batch_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        auto &info = m_pushed_info_control[static_cast<size_t>(batch_idx)];
+        if (info.value != setting) {
+            info.do_write = true;
+            info.value = setting;
+        }
+    }
+
+    double SysfsIOGroup::read_signal(const std::string &signal_name, int domain_type, int domain_idx)
+    {
+        std::string cname = check_request(__FUNCTION__, signal_name, "", domain_type, domain_idx);
```

Comment:
```suggestion
        std::string cname = check_request(__func__, signal_name, "", domain_type, domain_idx);
```
``__FUNCTION__`` is the old way: https://gcc.gnu.org/onlinedocs/gcc/Function-Names.html#:~:text=This%20name%20is%20the%20unadorned,with%20old%20versions%20of%20GCC.
We use ``__func__`` else ware.